internal/pkg/cli/deploy.go

```diff
@@ -244,31 +405,90 @@ func (o *deployOpts) Run() error {
 		return err
 	}
 
-	if err := o.maybeInitWkld(); err != nil {
+	deploymentOrderGroups, err := o.getDeploymentOrder()
+	if err != nil {
 		return err
 	}
 
-	if err := o.loadWkld(); err != nil {
-		return err
+	cmds := make([][]workloadCommand, len(deploymentOrderGroups))
+	// Do all our asking before executing deploy commands.
+	log.Infoln("Checking for all required information. We may ask you some questions.")
+	for i, deploymentGroup := range deploymentOrderGroups {
+		for _, workload := range deploymentGroup {
+			if err := o.maybeInitWkld(workload); err != nil {
+				return err
+			}
+			deployCmd, err := o.loadWkldCmd(workload)
+			if err != nil {
+				return err
+			}
+			cmds[i] = append(cmds[i], workloadCommand{
+				name: workload,
+				cmd:  deployCmd,
+			})
+			if err := deployCmd.Ask(); err != nil {
+				return fmt.Errorf("ask %s deploy: %w", o.wlType, err)
+			}
+			if err := deployCmd.Validate(); err != nil {
+				return fmt.Errorf("validate %s deploy: %w", o.wlType, err)
+			}
+		}
+	}
+
+	// Count number of deployments.
+	count := 0
+	for i := 0; i < len(cmds); i++ {
+		count += len(cmds[i])
 	}
-	if err := o.deployWkld.Execute(); err != nil {
-		return fmt.Errorf("execute %s deploy: %w", o.wlType, err)
+	log.Infof("Will deploy %d %s in the following order.\n", count, english.PluralWord(count, "workload", ""))
+	for i := 0; i < len(cmds); i++ {
+		names := ""
+		for _, cmd := range cmds[i] {
+			names += cmd.name + " "
+		}
+		log.Infof("%d. %s\n", i+1, names)
 	}
-	if err := o.deployWkld.RecommendActions(); err != nil {
-		return err
+
+	totalNumDeployed := 1
+	for _, deploymentGroup := range cmds {
+		// TODO parallelize this. Steps involve modifying the cmd to provide an option to
+		// disable the progress tracker, and to create several syncbuffers to hold spinners.
+		for g, cmd := range deploymentGroup {
+			if err := cmd.cmd.Execute(); err != nil {
+				return fmt.Errorf("execute deployment %d of %d in group %d: %w", totalNumDeployed, len(deploymentGroup), g+1, err)
+			}
+			if err := cmd.cmd.RecommendActions(); err != nil {
+				return err
+			}
+			totalNumDeployed++
+		}
 	}
+
 	return nil
 }
 
-func (o *deployOpts) askName() error {
-	if o.name != "" {
+func (o *deployOpts) askNames() error {
+	if o.workloadNames != nil || len(o.workloadNames) != 0 {
 		return nil
 	}
-	name, err := o.sel.Workload("Select a service or job in your workspace", "")
-	if err != nil {
-		return fmt.Errorf("select service or job: %w", err)
+
+	if !o.deployAllWorkloads {
```

Comment:
it might be easier to understand for 
```suggestion
	if o.deployAllWorkloads {
```